receptor/router.py

```
@@ -225,6 +226,6 @@ def next_hop(self, recipient):
         if expected_response and inner_envelope.message_type == 'directive':
             self.response_registry[inner_envelope.message_id] = dict(message_sent_time=inner_envelope.timestamp)
         if next_node_id == self.node_id:
-            await self.receptor.handle_message(msg)
+            asyncio.ensure_future(self.receptor.handle_message(msg))
```

Consider adding a callback on this `Future` for proper error logging/handling?

https://docs.python.org/3.6/library/asyncio-task.html#asyncio.Future.add_done_callback

---

I don't know if having a `Future` to catch the exception is worthwhile.

`handle_directive` seems to handle the exceptions as well as we can. `handle_response` *should* probably handle an exception as well as `handle_directive` does.

`handle_message` as long as the other things it calls does their best to handle an error then that's as good as it's going to be.

---

I could be 100% wrong, but I believe that any Exceptions raised during the execution of a task only get surfaced when the result is accessed, not when the Exception occurs. Since the code here isn't saving the future itself, it wouldn't otherwise access the Exception or surface it. Adding a callback to at least log an unhandled Exception might be wise. But again, could be 100% wrong as I often am with asyncio.

---

I think you are correct and we do need to add code to surface any exceptions. Or add a global exception handler (which @matburt) has been talking about).

---

I'd be alright just to log if we can't send a reply. It's fine if we swallow the exception (we don't want to surface it back to the command line anyway) as long as a log message is raised or a response it set.

Please don't use a callback here.

---

If not via a callback, how do you generally prefer to log (or generate a response message) from unhandled exceptions with asyncio Tasks?

---

For this case, catching exceptions inside of `handle_message` is preferable since that's where the context is more likely to exist.  I think `handle_directive` already does a good job, and `handle_response` could use a little attention.

---

Actually...

If the message type is not known, `handle_message` raises an exception itself.  Until that changes we'll have to handle that case.

---

So you'd want to wrap `handle_message` with a `try`/`except Exception` to ensure unhandled exceptions are impossible?

---

@matburt @jhjaggars @j00bar I wrapped `handle_message` as discussed.  Please let me know if you think this is the right approach.

---

That seems fine to me.

---

Meets my needs - thanks, Graham.